### ISSUE TITLE
[Feature request] Installation as local user (with admin privileges)

### ISSUE THREAD

**Issue by treysis** [notepad-plus-plus/notepad-plus-plus#15182]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description of the Issue

Currently, installation of Notepad++ requires access rights elevation (i.e. running as admin user) and sparks a UAC prompt. This requires that the admin first of all trusts the installer, but also prevents installation for users without admin rights. Running a portable version is not a good workaround, as it limits the capability of swift security updates and lacks explorer (right-click) integration.

There were some similar feature requests in the past, but I think it is time to revisit:
https://github.com/notepad-plus-plus/notepad-plus-plus/issues/388
https://github.com/notepad-plus-plus/notepad-plus-plus/issues/3425

I believe today the stance has changed so that app installations on a per-user level are a common thing. Especially since Microsoft Store with its app became more famous and moreover with the launch of Windows Package Manager (a.k.a. winget) and its success among users allowing a user install of Notepad++ is a valid request.

### Describe the solution you'd like.

I am not exactly sure what options NSIS offers, but I can think of the following solutions:
1. commandline switch
2. spawn UAC prompt and fall back to per-user install if no elevated privileges are granted

Generally, per-user installs should go into `%LocalAppData%\Notepad++`

### Debug Information

```shell
(none)
```

**Comment by jromal** [comment]
After many years using Notepad++ and recommending it to lots of people, I will stop using it.
The need of Admin Rights in Windows is absolutely a no-go. So many years, no answer, nothing done.
Sorry, but this is just an text editor. Admin rights for a text editor is really not in line with the current times.

**Comment by alankilborn** [comment]
@jromal 

Use portable version, placed in a folder location that requires no special "rights"?

**Comment by treysis** [comment]
@alankilborn no. Portable version is uncomfortable to update and most importantly, it lacks context menu integration!
Somebody 1 or 2 years ago published a version to Microsoft Store and it worked great. Unfortunately, Notepad++ had it removed from the store. Now that I got a new computer from work I can't install it anymore from the store and thus can't use Notepad++ anymore. And I don't want to ask IT department anytime there's an update to Notepad++.

**Comment by xomx** [comment]
https://nsis.sourceforge.io/Reference/RequestExecutionLevel

Using `none` or `user` in the N++ installer source NSIS-script could do.
But the problem probably will be the possible installation of the NppShell extension (I did not test, maybe it could be now per-user too, IDK).

I will try to play with it but I am afraid that any such attempt could only end up with making an additional version of the installer published, which is a no-go.

> Portable version is uncomfortable to update

Well, you want something special, so it needs some effort from your side. It is doable e.g. via PythonScript ([Upgrading notepad++ with one click.](https://community.notepad-plus-plus.org/post/95098) ) or simply using a package manager service like the MS [WinGet](https://winget.run/pkg/Notepad++/Notepad++).

> Portable version ... it lacks context menu integration!

You can extract it (e.g. 7-Zip app can unpack the NSIS-installers) from the installer (look there for the `contextMenu` subfolder) and register it manually (`regsvr32 NppShell.dll` at cmdline) even for the portable version (I do not recommend that but as you wish).

**Comment by jromal** [comment]
> @alankilborn no. Portable version is uncomfortable to update and most importantly, it lacks context menu integration! Somebody 1 or 2 years ago published a version to Microsoft Store and it worked great. Unfortunately, Notepad++ had it removed from the store. Now that I got a new computer from work I can't install it anymore from the store and thus can't use Notepad++ anymore. And I don't want to ask IT department anytime there's an update to Notepad++.
> 
> @jromal can you suggest any good alternative?

I fully agree. It is not the same a portable version (without updates) and the real thing.

Currently I am tending to use VSCode, even if it is too big for most of the things and it is "monitored" by Microsoft. 
But at least you can install it anywhere without asking the IT department.

**Comment by jromal** [comment]
I do not want alternative ways to install. I want to know:
- why does it need admin rights for installation? 
- What are the features that require it and cannot be implemented otherwise?
- Why there is no option to have a version without those features that does not required elevated privileges?

It looks like a program that has lost the contact with the reality.

**Comment by treysis** [comment]
> * What are the features that require it and cannot be implemented otherwise?

None. As I mentioned, there used to be an (unofficial) version published on Microsoft Store 1 or 2 years ago and it was working flawlessly.

> I will try to play with it but I am afraid that any such attempt could only end up with making an additional version of the installer published, which is a no-go.

I don't think that would be an issue. There are several apps on the winget repo that have a separate version for per-user installs. Although I agree it would be cleaner to manage this within one single installer.

Typically I know this from installations where the script asks "this user only" or "all users". I don't know if the N++ installer supports this.



> You can extract it (e.g. 7-Zip app can unpack the NSIS-installers) from the installer (look there for the `contextMenu` subfolder) and register it manually (`regsvr32 NppShell.dll` at cmdline) even for the portable version (I do not recommend that but as you wish).

Why don't you recommend this?

I don't know if this will just work on no-admin accounts but I will give it a try! In general this is possible, though. Other apps make use of it. E.g. NanaZip - a fork of 7zip on the MS store - has perfect context menu integration on user-only installs without admin.

**Comment by treysis** [comment]
I tried the manual registration of shell extensions but it didn't work. I also have no idea how the registration should know to where I unzipped N++.

**Comment by xomx** [comment]
@treysis 

> I also have no idea how the registration should know to where I unzipped N++.

Ah, I forgot to add:
- put the `contextMenu` subfolder with the NppShell stuff to your portable main N++ folder ([relevant NppShell source code](https://github.com/notepad-plus-plus/nppShell/blob/7b715ada2749321f0bf8fc08a5587ea2d10ab9cf/PathHelper.cpp#L16-L20)): 
  ![nppshell-contextMenu-subfolder-placement](https://github.com/user-attachments/assets/b71729ac-4e92-44fa-98cb-8a1c2432822b)

- register/unregister it (but for that you will need the admin rights, so no per-user shell extension for now, as I expected):
  ![nppshell-contextMenu-subfolder-content](https://github.com/user-attachments/assets/71d96c69-1091-4903-b00f-cab1e17d0be6)


> Why don't you recommend this?

It is complicated and I do not have time to explain, sorry.

**Comment by treysis** [comment]
@xomx Ok, got it working. The important part was to run the command from the parent folder of the notepad++.exe: `regsvr32.exe .\contextMenu\NppShell.dll`). Doesn't require admin rights. I now have it in my right click context menu.

While I still think N++ should be converted to user-only installation by default, this is certainly a neat workaround that will make it easier at least for me.

If I can craft an installer that does a user-only install, would you be interested in publishing it?

**Comment by xomx** [comment]
@treysis 

> Doesn't require admin rights. I now have it in my right click context menu.

Ok, I can confirm that the N++ shell context menu is there after registering W/O admin rights (I have to restart my filemanager app used after that registration to see the menu, logoff/logon can help too in such a situation).

But there is a problem - when you run the cmdline `regsvr32 NppShell.dll`, you should obtain this message from the regsvr32 (if you do not use the `/s` - silent mode switch):

![nppshell-regsvr32-msg](https://github.com/user-attachments/assets/df28199f-de1e-4b19-b6fe-aae3e69747b6)

But this is not the case without the admin rights - I am not getting such msg. The registration is done somehow & works but there remains a "zombie" `regsvr32.exe` process visible in the TaskManager without any active thread but still with some resources allocated. This looks to me as a bug in the NppShell project.

> If I can craft an installer that does a user-only install, would you be interested in publishing it?

This is not a question for me, but for this project maintainer.

**Comment by xomx** [comment]
> If an Admin doesnt want this to be allowed then they should be able to turn off and manage the plugins manually. But there should be the option.

https://github.com/notepad-plus-plus/notepad-plus-plus/issues/14277#issuecomment-1782563875

**Comment by ependergrass** [comment]
This is causing issues with my new "no admin access" work laptop sadly.. I was able to get workarounds for almost every other application except this one, including availability in MS store or the installer working properly and not requiring admin access. Hope we can get this soon!

**Comment by xomx** [comment]
@ependergrass 
> get workarounds

Meanwhile (I'm currently busy elsewhere and don't see any opportunity to devote myself to this in the near future), using the portable N++ versions (e.g. [npp.8.8.3.portable.x64.zip](https://github.com/notepad-plus-plus/notepad-plus-plus/releases/download/v8.8.3/npp.8.8.3.portable.x64.zip)) instead of the installed ones isn't a solution for you?

**Comment by treysis** [comment]
@xomx does it have context menu integration, or not? That's the reason why I mostly stopped using Notepad++.

**Comment by victorel-petrovich** [comment]
@treysis , I'm curious, what's the big deal with context-menu integration, that "Edit with N++" thing? Once you use "Open with..." and associate to use  N++ to open a particular filetype, then just double-click. How often do you need to open a file not set to be default w/ N++? In those cases, "Open with" is only a few more clicks/press keys away from "Edit w/ N++". 
Besides, that integration is still possible to do for Portable; just a bit more manual setup.

**Comment by treysis** [comment]
@victorel-petrovich Yes. Unfortunately, in our corp environment, I don't have too much control over file type associtations and they seem to get reset from time to time. Also, most of the time even N++ is overkill and I prefer the tab-less MS Notepad. Also, sometimes I have special software project files that I might want to edit but by default should launch the special application.

My environment might be a special case. Not sure. However, I don't have control over those things, whereas it would be relatively easy to integrate those settings into N++.

For now, I have switched to using VS Code instead, which doesn't have these issues that N++ does (and it was also a simple non-admin user install).

**Comment by victorel-petrovich** [comment]
@treysis shouldn't the title be "Installation as local user (**without** admin privileges)" ? 

@Will-EML 
> Basically there needs to be the option that Admins can tun on or off for user to be able to install plugins in a user context.

IMHO, that's way too much power over users. What goes into "user context" should be user's business. Also futile, because  users can install the portable and could all the plugins they want anyway.

**Comment by DraftingDreamer** [comment]
Hi everyone — after reading the discussion here (and the now-closed #3425), I took a swing at implementing this and have a working PoC on a branch.

I tried to stay close to the unified-installer / dynamic-elevation approach @xomx outlined earlier in this thread: single installer, no UAC at startup, only elevate if the user actually picks the all-users mode.

**Branch / diff (5 NSIS files, +261 / -90, no C++ changes):**
https://github.com/notepad-plus-plus/notepad-plus-plus/compare/master...DraftingDreamer:notepad-plus-plus:perUserInstall_20260615

**Unsigned PoC installer for quick hands-on (x64 / x86 / arm64):**
https://github.com/DraftingDreamer/notepad-plus-plus/releases/tag/peruser-poc-v8.9.6.4

**Design summary**

| | Behavior |
|---|---|
| Startup | `RequestExecutionLevel user` — no UAC for any account |
| New page | "Choose Installation Type" after License (⦿ all users / ⦿ just for me; both always selectable) |
| Per-user default | `$LOCALAPPDATA\Notepad++`, registry under HKCU via `SHCTX` |
| All-users default | unchanged: `$PROGRAMFILES64\Notepad++`, HKLM |
| All-users without an admin token | `ExecShell "runas"` relaunch with `/AllUsers` |
| NppShell context-menu | `SHCTX\Software\Classes` (HKCR for all-users / `HKCU\Software\Classes` for per-user — no admin needed); MSIX path untouched |
| Uninstaller | reads `InstallMode` marker in `un.onInit`; cleans the right hive + shortcut scope |
| `doLocalConf.xml` portable mode | untouched |
| C++ (`notepad++.exe`) | **no changes** — `Parameters.cpp` already picks `%APPDATA%` whenever the exe is not under Program Files |

**Tested on x64** (silent and GUI):
- Per-user install → files in `%LOCALAPPDATA%`, HKCU keys, HKLM untouched
- All-users install → identical to current behavior, single UAC prompt after the choice
- Silent uninstall → fully clean, no residue in HKCU/files/shortcuts
- Coexists correctly with an existing all-users install on the same machine

x86 and arm64 build cleanly from the same script; I don't have those hardware to run the installer on.

**Known follow-ups I'd address in a real PR** (intentionally scoped out of this PoC):

- `uninstall.exe` filename trips Windows installer-detection and asks for UAC even with an `asInvoker` manifest — needs either a rename or a manifest override
- x86 NppShell registration still calls `regsvr32 /s NppShell.dll`; the manual `SHCTX\Software\Classes\CLSID\{...}` writes already cover the main registration, but the `DllRegisterServer` side effects deserve a cleaner per-user path
- MSI installer (`installer/msi/*.wxs`) not in scope

**Not opening a PR**, per CONTRIBUTING.md — this is enhancement-class work and needs the issue to be `Accepted` first. Posting the branch + binaries so it's reviewable, and so @xomx (or anyone else) can pick it up, fork it, or push back on the design before there's any commitment to a PR. Happy to iterate if there's interest.

**Comment by tiljes-st** [comment]
Wow, very cool! I had success installing it and it looks fine so far! Unfortunately, some features in Notepad++ stupidly require admin access (e.g. proxy setting...why?), but that cannot be solved by adapting the installer.

I'll have to see how well it works in a Microsoft FSLogix environment.

**Comment by xomx** [comment]
> some features in Notepad++ stupidly require admin access (e.g. proxy setting...why?)

Fixed by https://github.com/notepad-plus-plus/notepad-plus-plus/commit/bb7d108799bdbc899ca87e7481f42c2ad4f6fa5d for the upcoming v8.9.8.

**Comment by DraftingDreamer** [comment]
@xomx — following up on your analysis from last May:

> I think I should be able to re-code the unified installer so that it doesn't automatically require the admin rights at the very start, but then, if the user chooses to install in to a protected storage location, to fire-up dynamically the UAC-prompt and relaunch.

That is what the PoC I posted in June does, and it works. It follows your outline: `RequestExecutionLevel user`, so there is no UAC prompt at startup, and the all-users path relaunches itself elevated through `ExecShell "runas"` only when the user actually picks that mode.

It also answers the concern you raised earlier in this thread:

> I am afraid that any such attempt could only end up with making an additional version of the installer published, which is a no-go.

There is no additional installer. It is the same `nppSetup.nsi` producing the same three binaries; the only addition is a "Choose Installation Type" page, and the all-users path behaves exactly as it does today. Registry writes go through `SHCTX`, so HKLM vs. HKCU follows from the selected mode instead of being a second code path. No C++ changes are needed — `Parameters.cpp` already falls back to `%APPDATA%` whenever `notepad++.exe` is not under Program Files.

The design table, the branch compare and the test builds are in my June comment above. @tiljes-st installed it successfully and reported back here. The five NSIS files the change touches have not been modified upstream since the branch point, so the change is still current.

Why I think it is worth doing: on a locked-down corporate machine the user has no admin rights but still needs Notepad++. The portable build is the answer today, but keeping it updated and getting the shell extension registered both require manual workarounds that most users in that situation cannot apply.

Per CONTRIBUTING.md — *"For issues that have not been accepted, you may request to be assigned to that issue"* — I would like to request assignment of this issue, and I will open the PR once it is accepted.

If I take it, these are the two loose ends the PoC deliberately left open, and I would close them in the PR rather than leave them behind:

- `uninstall.exe` trips Windows' installer detection because of its file name; it needs either a rename or an `asInvoker` manifest.
- The x86 path still registers NppShell through `regsvr32 /s`; the per-user route should be cleaner than that.

The MSI packages are deliberately out of scope: as @donho put it in #17169, the MSI is for enterprise IT deployment only, which is a different delivery path from a user installing Notepad++ for themselves.

And if you would rather do this yourself, that is completely fine — take whatever is useful from the branch. I would just like to know either way, so that the PoC does not sit here unused.

**Comment by xomx** [comment]
@DraftingDreamer 
I'm glad you took on this and are trying to implement it yourself. My spare time is unfortunately low but if I find a time, I'll review this very useful new possibility for the N++ installer. I agree that having an option for a local user installation is important these days.
